src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -479,9 +479,9 @@ internal struct HeaderEntry<T>
         /// If the symbol of the extracted declaration is not an unqualified symbol,
         /// verifies that it corresponds instead to an invalid symbol and returns null unless the keepInvalid parameter has been set to a string value.
         /// If the keepInvalid parameter has been set to a (non-null) string, uses that string as the SymbolName for the returned HeaderEntry instance.
-        /// Throws an ArgumentException if this verification fails as well.
-        /// Throws an ArgumentException if the extracted declaration is Null.
         /// </summary>
+        /// <exception cref="ArgumentException">Thrown if the symbol of the extracted declaration is not an unqualified symbol and doesn't instead correspond to an invalid symbol.</exception>
+        /// <exception cref="ArgumentException">Thrown if the extracted declaration is Null.</exception>
```

Comment:
What do you think about combining two exception tags into one tag if they have the same exception type? I think this would be more consistent with some of the other doc comments that say "thrown if X, or Y, or Z."

Reply:
You know, I hadn't thought about that. I just now spent a little time digging.

The Sandcastle Help File Builder documentation seems to indicate that [only a single tag should be present per exception type](https://github.com/EWSoftware/SHFB/blob/ea37113a5aa5e6d998213d37f43cc4f49acac2a7/Documentation/XMLCommentsGuide/Content/SectionElements/exception.aml#L23), though this documentation hasn't been updated since 2012.

On the flip side, I found an MS example where [multiple entries are used](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/src/CoreLib/System/Threading/Tasks/TaskCompletionSource.cs#L178). In this case, whatever tool that generated the API docs seems to have [automatically joined these entries](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1.trysetexception?view=netcore-2.1#System_Threading_Tasks_TaskCompletionSource_1_TrySetException_System_Collections_Generic_IEnumerable_System_Exception__) with `<para>- or -</para>` between each.

Personally, I think I'm in favor of combining, but perhaps with some structure for different causes. Maybe something like:

```xml
<exception cref="ArgumentException">
  Causes:
  <list type="bullet">
    <item>
      <description>The symbol of the extracted declaration is not an unqualified symbol and doesn't instead correspond to an invalid symbol.</description>
    </item>
    <item>
      <description>The extracted declaration is Null.</description>
    </item>
  </list>
</exception>
```

Which when generated with DocFX looks like this:
<img width="706" alt="Screen Shot 2020-11-15 at 2 42 21 AM" src="https://user-images.githubusercontent.com/1374024/99179482-36e44900-26ec-11eb-8dd0-f85400da771c.png">

I think for nearly identical causes (e.g. an exception is thrown if any of a few parameters are null), a single sentence like "thrown if X, or Y, or Z" is plenty readable and fine as is.

So I would propose:
- For any exception tags with the same `cref`, combine into a single tag using a list format.
- Leave any single sentences like "thrown if X, or Y, or Z" alone, unless they're obviously overloaded with logically separate causes.

I think that's sort of what's done at least for [this example](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.appendalllines?view=net-5.0), where `ArgumentNullException` uses a sentence and `UnauthorizedAccessException` uses a list (kinda...).

Reply:
Oh, interesting, I didn't realize the .NET documentation used multiple `exception` tags like that. I think as long as we're consistent in how we apply the distinction, like you mentioned (closely related conditions vs. unrelated conditions), do you see any downside to using separate tags to represent an "or" of unrelated conditions? It looks less verbose in the source code compared to the list, and IMO the list is a little ambiguous about whether the list items are connected by "or" or "and," whereas with separate tags it makes more sense to me that the conditions would be a list of separate possibilities (so "or" instead of "and").

Reply:
> do you see any downside to using separate tags to represent an "or" of unrelated conditions?

I think my only concern is how tools will handle it when rendering. Unfortunately, it looks like DocFX creates a new table row for every exception tag, even if they share the same `cref` (I'm not sure what .NET docs use). This would be hard to read in API docs IMO. I wonder if you might be able to test this out in Rider?

> It looks less verbose in the source code compared to the list

Completely agree. I rather like the way the separate tags approach looks in the source.

> the list is a little ambiguous about whether the list items are connected by "or" or "and,"

Darn. I thought maybe I'd found a more elegant solution than separating conditions with `"- or -"` by heading the list with `Causes:`.

We could always just go with the "- or -" thing. It's actually fairly readable in the source IMO.

```c#
/// <exception cref="ArgumentException">
///     <para>The symbol of the extracted declaration is not an unqualified symbol and doesn't instead correspond to an invalid symbol.</para>
///     <para>- or -</para>
///     <para>The extracted declaration is Null.</para>
/// </exception>
```

Here's a render in DocFX next to the list approach:

<img width="710" alt="Screen Shot 2020-11-17 at 12 47 26 AM" src="https://user-images.githubusercontent.com/1374024/99351488-ad10b900-286e-11eb-84e8-dbd895f9b720.png">

Reply:
I think that I would prefer to pick a source representation that has the most flexibility in how it's rendered, since the documentation generator is easier to change than re-doing all of the doc comments.

The multiple `exception` tags seem to encode more information in a machine-readable format than the list or paragraph versions. I think it'd be easier to have the docs generator transform the multiple `exception` tags into the format that we want, than it is to transform the list or paragraph versions into a different format if we ever decide to change the way the rendered docs look like. It also has the benefit of being simpler to write in the source code.

We don't currently generate docs for the Q# compiler, but if/when we do I think we can address the multiple exceptions of the same type problem as being a problem with the generator/template we use, instead of a problem with the source. Does that seem reasonable?

Also, I tried it in Rider and it shows each exception tag separately (like DocFX) - that's probably harder to change than the docs generator template, but personally I think it's not that much of a readability issue to work around it in the source code.

Reply:
> The multiple exception tags seem to encode more information in a machine-readable format

+1

> we can address the multiple exceptions of the same type problem as being a problem with the generator/template we use

That's a great idea! It'd be pretty easy to add a post-processing step for the full XML files that the C# compiler spits out. This also means less work for me to get this PR ready :D